Feature/printer settings dialog improvements

NotExpectedYet opened the pull request:
Printer Settings Class refactor. 

Cleaned up a lot of code so far, still some things left to do. 

Using this as the base to start the work on #532 #531

 - [x] Replace functions that load the settings modal. 
 - [x] Refactor the code to load everything under it's own static function
 - [x] Add in try/catches and respond the error to the client if a problem arises
 - [x] Create a new octofarm client file and improve the functions with error responses in a single place
 - [x] Add in new api endpoint to trigger and update for a printers settings from octoprint
 - [x] Make settings panel trigger the update from OctoFarm for octoprint's latest settings. 
 - [x] Replace global try/catch with multiple so 1 panel failing to load won't break the rest
 - [x] Put back in offline checks into panels
 - [x] Put back in offline message
 - [x] Add in check for port not available in drop down
 - [x] Add back in listener for save function
 - [x] Refactor the save function
 - [x] Fix save button event listener not applying
 - [x] Add kill button event listener
 - [x] Add test for endpoint
 - [x] Add changelog
 - [x] Add in port refresh
 - [x] Combine response into 1 alert.
 - [x] Add page refresh after save
 - [x] Clean up the extra calls on settings saving. Client refresh will make these redundant.

davidzwa commented on server_src/routes/printers.js:
```diff
@@ -102,7 +102,7 @@ router.post("/updateSettings", ensureAuthenticated, async (req, res) => {
   logger.info("Update printers request: ", settings);
   const updateSettings = await Runner.updateSettings(settings);
-
+  res.sendStatus(200);
```
Why should we set a status if default is 200? Im a lazy .NET dev, all my status codes are automagically set XD

NotExpectedYet replied:
I think this can be removed, I was chasing the settings issue and thought it was here. Will test and remove. Thought it was sending a 204 for some reason.

davidzwa replied:
Nothing wrong with a 204: just means the body is empty.

NotExpectedYet replied:
I want a body, I specifically look for 204's in the new OctoFarm client.

davidzwa commented on views/printerManagement.ejs:
```diff
@@ -564,27 +564,27 @@
                         <div class="col-12 col-sm-12 col-md-12 col-lg-2 col-xl-2">
                             <div class="list-group" id="list-tab" role="tablist">
-                                <a class="list-group-item list-group-item-action active" id="printer-connection-btn"
+                                <a class="list-group-item list-group-item-action active notyet" id="printer-connection-btn"
```
What does this class name do?

NotExpectedYet replied:
disables an a tag

davidzwa reviewed (changes requested):
The PR does not fix #532, so my proposal is to remove that from the linked issues. The reason for this is because I have a more effective solution for it in mind for Printer Clean (for which it would be great if #406 is merged first, that would help me not having to pile a PR on top of another PR like file and job clean).

If you fix the event listener, remove the issue from linked issues I think this PR is good to go!

Dont forget:
- changelog
- Jest tests

NotExpectedYet commented:
I'll undo the mystery machine 🤣

NotExpectedYet commented:
Done the redirect mock, but not a clue with mocking passport.js at moment will look later when I have more time.

NotExpectedYet commented on client_src/js/lib/functions/ui.js:
```diff
@@ -89,4 +89,17 @@ export default class UI {
       });
     });
   }
+  static removeLoaderFromElementInnerHTML(element) {
+    if (element.innerHTML.includes("spinner")) {
+      element.innerHTML = element.innerHTML.replace(
+        '<i class="fas fa-spinner fa-spin"></i>',
+        ""
+      );
+    }
+  }
+  static addLoaderToElementsInnerHTML(element) {
+    if (!element.innerHTML.includes("spinner")) {
+      element.innerHTML += ' <i class="fas fa-spinner fa-spin"></i>';
+    }
+  }
```
Just tow new UI functions that I use quite a lot. Rest of client can start utilising these more but will migrate in another PR.

NotExpectedYet replied:
The remove from loader was also causing the listener to disappear which is something I'll have to keep in mind. Initially there was no check if "spinner" was included in the innerHTML so it was replacing nothing basically. That was causing the listener to disappear which I had no idea was an effect of that.

NotExpectedYet commented on client_src/js/lib/octofarm_client.js:
```diff
@@ -0,0 +1,42 @@
+export default class OctoFarmClient {
+  static async get(item) {
+    const url = `/${item}`;
+    const get = await fetch(url, {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+      },
+    });
+    if (get.status === 400) {
+      throw new Error(
+        `Malformed request! Status: ${get.status} - ${get.statusText}`
+      );
+    } else if (get.status === 503) {
+      throw new Error("Error contacting server, is it alive?");
+    } else if (get.status === 204) {
+      return;
+    }
+    return await get.json();
+  }
+  static async post(item, data) {
+    const url = `/${item}`;
+    const post = await fetch(url, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(data),
+    });
+    if (post.status === 400) {
+      throw new Error(
+        `Malformed request! Status: ${post.status} - ${post.statusText}`
+      );
+    } else if (post.status === 503) {
```
Created a new OctoFarm client that does the status checks and then passes back json if available. Simplifying the code required to utilising it and stopping all the repeated status checks elseware. Plan to migrate rest of code in another PR but left the old one in for now.

NotExpectedYet commented on client_src/js/printerManagementRunner.js:
```diff
@@ -3,7 +3,7 @@ import OctoPrintClient from "./lib/octoprint.js";
 import UI from "./lib/functions/ui.js";
 import PrinterManager from "./lib/modules/printerManager.js";
-import PrinterSettings from "./lib/modules/printerSettings.js";
+import { updatePrinterSettingsModal } from "./lib/modules/printerSettings.js";
```
Rather than importing the full class I decided to just create a function that applied the state and utilised the statics in the file.

NotExpectedYet commented on client_src/js/printerManagementRunner.js:
```diff
@@ -94,11 +91,23 @@ if (window.Worker) {
   } catch (e) {
-    console.log(e);
+    UI.createAlert(
+      "error",
+      "Sorry web workers are not supported in your browser, please check the supported browser list.",
+      0
+    );
+    console.error(
+      `Couldn't create the web worker, please log a github issue... <br> ${e}`
+    );
   }
 } else {
   // Sorry! No Web Worker support..
-  console.log("Web workers not available... sorry!");
+  UI.createAlert(
+    "error",
+    "Sorry web workers are not supported in your browser, please check the supported browser list.",
+    0
+  );
+  console.error(`Web workers not available... sorry! <br> ${e}`);
```
Updated these with a UI alert rather than hiding in console.

NotExpectedYet commented on client_src/js/printerManagementRunner.js:
```diff
@@ -3321,7 +3329,7 @@ class dashUpdate {
               .getElementById(`printerStatistics-${printer._id}`)
               .addEventListener("click", async (e) => {
-                PrinterLogs.loadStatistics(printer._id);
+                await PrinterLogs.loadStatistics(printer._id);
```
Missing await

NotExpectedYet commented on server_src/lib/dataFunctions/printerClean.js:
```diff
@@ -61,8 +61,15 @@ class PrinterClean {
    */
-  static returnPrintersInformation() {
-    return printersInformation;
+  static returnPrintersInformation(id) {
+    if (!id) {
+      return printersInformation;
+    } else {
+      const index = _.findIndex(printersInformation, function (o) {
+        return o._id == id;
+      });
+      return printersInformation[index];
+    }
```
Updated this function to work with and without a specific printer. No id will return the entire list, and id provided will pull out the requested printer. This will simplify the code elseware a lot with the check done here. Follow up PR to update to utilise this more but will await the cache finishing for printers.

davidzwa replied:
Im not a fan of this code style as the function becomes indeterministic. Simpler functions have simpler bugs.
Usually I create a 'getById' and a 'list' variant to be 100% clear, also because this is the CRUD approach (get, list are separated). Can you point to the code which becomes much simpler, help me see the impact a bit?

Btw, of course sometimes ease of use is really nice. I fully understand that 😉. Just wanted to let you know that in such cases either the code which uses this function could benefit from unit tests to make absolutely sure the `id` parameter is not left null/false/empty/0/NaN. If I see those test, you would not have heard me ring a bell as I would know you'd tested it and your thus in full control. 
Those tests are probably high in count, so splitting it up is: 1) less impactful 2) really clear 3) less code to cover in tests

davidzwa replied:
![image](https://user-images.githubusercontent.com/6005355/119563337-af1c7d00-bda7-11eb-9a1c-dbf4495dd811.png)
Changing the function to 2 types doesnt seem like it will impact the code.

NotExpectedYet commented on server_src/routes/printers.js:
```diff
@@ -9,10 +9,14 @@ const { Runner } = runner;
 
 const { PrinterClean } = printerClean;
 
+// Doesn't returns undefined, note to use is incorrect...
+// const { returnPrintersInformation } = require("../cache/printer.cache.js");
+
```
Just a comment when I was testing, can be removed when cache is implemented.

NotExpectedYet commented on server_src/routes/printers.js:
```diff
@@ -98,14 +102,12 @@ router.post("/feedChange", ensureAuthenticated, async (req, res) => {
   res.send({ status: updateSettings.status, printer: updateSettings.printer });
 });
-router.get("/killPowerSettings/:id", ensureAuthenticated, async (req, res) => {
+router.post("/killPowerSettings/:id", ensureAuthenticated, async (req, res) => {
```
Updated to a post command as backend is changed

davidzwa replied:
I dont understand this comment as this is the backend we're looking at?

NotExpectedYet commented on server_src/routes/printers.js:
```diff
@@ -125,25 +127,26 @@ router.get("/groups", ensureAuthenticated, async (req, res) => {
 });
 router.post("/printerInfo", ensureAuthenticated, async (req, res) => {
   const id = req.body.i;
+  const returnPrinter = await PrinterClean.returnPrintersInformation(id);
+  res.send(returnPrinter);
+});
```
Simplified the code with the new allowance of an ID check within return printers information

NotExpectedYet commented on server_src/routes/printers.js:
```diff
@@ -125,25 +127,26 @@ router.get("/groups", ensureAuthenticated, async (req, res) => {
-      systemChecks: printers[index].systemChecks,
-    };
-    res.send(returnPrinter);
+router.post("/updatePrinterSettings", ensureAuthenticated, async (req, res) => {
+  const id = req.body.i;
+  if (!id) {
+    logger.error("Printer Settings: No ID key was provided");
+    res.statusMessage = "No ID key was provided";
+    res.sendStatus(400);
+    return;
+  }
+  try {
+    await Runner.getLatestOctoPrintSettingsValues(id);
+    let printerInformation = PrinterClean.returnPrintersInformation(id);
+    res.send(printerInformation);
+  } catch (e) {
+    logger.error(`The server couldn't update your printer settings! ${e}`);
+    res.statusMessage = `The server couldn't update your printer settings! ${e}`;
+    res.sendStatus(500);
   }
 });
```
New endpoint to update printer settings, must provide an ID for it. Sends information to client when no id provided with status message. I would like to utilise these status messages a lot more as it will help people using the client in the future to actually get an error response back rather than just a status code.

NotExpectedYet commented on server_src/runners/state.js:
```diff
@@ -3000,6 +2998,23 @@ class Runner {
         return false;
       });
   }
+  // Patch for updating OctoPrint's settings for now until re-work of printer cache with state.js.
+  static async getLatestOctoPrintSettingsValues(id) {
+    const index = _.findIndex(farmPrinters, function (o) {
+      return o._id == id;
+    });
+    // This is why the settings we're not updating! Forgot that connection options and preferences come in state, not settings/system.
+    await Runner.getState(id);
+    // Update the printers cached settings from OctoPrint
+    await Runner.getSettings(id);
+    // Update the printers cached system settings from OctoPrint
+    await Runner.getSystem(id);
+    // Re-generate the printer clean information - This is just cautionary, my tests showed it wasn't needed.
+    await PrinterClean.generate(
+      farmPrinters[index],
+      systemSettings.filamentManager
+    );
+  }
```
New function for state. Just runs through the required calls to OctoPrint that are needed to update the printer settings on OctoFarms side. The extra generate wasn't needed but I'd rather keep it just encase there are timing issues server side.

NotExpectedYet replied:
Can be cleaned up more once the refactor of state JS has happened.

NotExpectedYet commented on test/api/printers-check.test.js:
```diff
@@ -0,0 +1,50 @@
+  return app;
+}
+
+describe("Printer Update Endpoint", () => {
+  it("should return 302 redirect when no user is logged in", async function (done) {
+    process.env.npm_package_version = require("../../package.json").version;
+    process.env.testlatest_package_version = require("../../package.json").version;
+    app = await getOrCreateApp();
+    const res = await request(app)
+      .post("/printers/updatePrinterSettings")
+      .send();
+    expect(res.statusCode).toEqual(302);
+    done();
+  }, 10000);
```
Initial test for redirect, but will need to update once I figure out how to do passport auth

NotExpectedYet commented on views/printerManagement.ejs:
```diff
@@ -564,32 +564,31 @@
-                                            class="fas fa-tags"></i> Profile</a>
-                                <a class="list-group-item list-group-item-action" id="printer-power-btn"
+                                            class="fas fa-tags"></i> Profile <i class="fas fa-spinner fa-spin"></i></a>
+                                <a class="list-group-item list-group-item-action notyet" id="printer-power-btn"
```
notyet is a specific class I created to disable the action on a links. It's removed from here once the section has loaded. It's handy if you want to disable an "a" link as they don't work with disabled as buttons do.

NotExpectedYet commented on client_src/js/lib/modules/printerSettings.js:
```diff
@@ -1,1325 +1,1403 @@
-// let controlDropDown = false;
-
-let currentPrinter = null;
+let currentPrinterIndex;
+let printerOnline;
+let currentPrinter;
+let hasErrorNotificationBeenTriggered = false;
+let pageElements;
```
global elements to store page information. 
CurrentPrinterIndex = used when the sse event fires again so we don't have to keep checking it. Number
printerOnline = boolean for status check on printers state. This is relevant to the host status. Not the printer status
currentPrinter = Object containing the current printer
hasErrorNotificationBeenTriggered = boolean that get's updated so the UI alert doesn't keep firing for sse event
pageElements = holder for the page elements

NotExpectedYet commented on client_src/js/lib/modules/printerSettings.js:
```diff
@@ -1,1325 +1,1403 @@
-  //   controlDropDown = false;
-  // }
-});
 $("#connectionModal").on("hidden.bs.modal", function (e) {
   if (document.getElementById("connectionAction")) {
     document.getElementById("connectionAction").remove();
   }
 });
```
Listens for the modal closing and removes the connectionAction element so it can be replaced.

NotExpectedYet commented on client_src/js/lib/modules/printerSettings.js:
```diff
@@ -1,1325 +1,1403 @@
-        UI.clearSelect("ps");
-        const printerProfileBtn = document.getElementById(
-          "printer-profile-btn"
+async function updatePrinterSettingsModal(printersInformation, printerID) {
```
This function is fired initially from the API call. It receives the printer ID and updates the rest of the page with that specific printer. 
It's then fired again from the SSE event, without a printer ID this time. The variables above will be filled by this time with the specific printers ID so we can continue updating the live elements. It's mainly the printer status stuff at the bottom of the page.

NotExpectedYet commented on client_src/js/lib/modules/printerSettings.js:
```diff
@@ -1,1325 +1,1403 @@
+
+    const alertsTable = document.getElementById("printerAltersTableBody");
+    alertsTable.innerHTML = "";
+    printerScripts.map(async (s) => {
+      alertsTable.insertAdjacentHTML(
+        "beforeend",
+        `
+            <tr>
+              <td>
+              <form class="was-validated">
+                    <div class="custom-control custom-checkbox mb-3">
+                          <input type="checkbox" class="custom-control-input" id="activePrinter-${s._id}" required="">
+                          <label class="custom-control-label" for="activePrinter-${s._id}"></label>
+                      </div>
+                  </form>
+              </td>
+              <td>
+              <select class="custom-select" id="triggerPrinter-${s._id}" disabled>
+               </select>
+                     </td>
+              <td>${s.scriptLocation} </td>
+              <td>${s.message}</td>
+            </tr>
+
+        `
+      );
+      document.getElementById(`activePrinter-${s._id}`).checked = s.active;
+      const triggerSelect = document.getElementById(`triggerPrinter-${s._id}`);
+      triggerSelect.innerHTML = await Script.alertsDrop();
+      triggerSelect.value = s.trigger;
+    });
```
Just trying out .map here rather than for each. Depends what your doing really but this can be faster and more resilient.

NotExpectedYet commented on client_src/js/lib/modules/printerSettings.js:
```diff
@@ -1,1325 +1,1403 @@
-    elements.mainPage.host.className = `btn btn-${printer.hostState.colour.name} mb-1 btn-block`;
-    elements.mainPage.socket.innerHTML = `<b>WebSocket Status</b><br>${printer.webSocketState.desc}`;
-    elements.mainPage.socket.className = `btn btn-${printer.webSocketState.colour} mb-1 btn-block`;
+  static disablePanel(element) {
+    if (element.classList.contains("notyet")) {
+      element.classList.remove("notyet");
+    }
+  }
```
I may move these to UI once more places want to utilise them. Handy functions.

NotExpectedYet commented on client_src/js/lib/modules/printerSettings.js:
```diff
@@ -1,1325 +1,1403 @@
+      pageElements.menu.printerProfileBtn.disabled = true;
+      pageElements.menu.printerGcodeBtn.disabled = true;
+      pageElements.menu.printerOtherSettings.disabled = true;
+
+      if (!pageElements.menu.printerProfileBtn.classList.contains("notyet")) {
+        pageElements.menu.printerProfileBtn.classList.add("notyet");
       }
-      if (!printerGcodeBtn.classList.contains("notyet")) {
-        printerGcodeBtn.classList.add("notyet");
+      if (!pageElements.menu.printerGcodeBtn.classList.contains("notyet")) {
+        pageElements.menu.printerGcodeBtn.classList.add("notyet");
       }
-      if (!printerOtherSettings.classList.contains("notyet")) {
-        printerOtherSettings.classList.add("notyet");
+      if (
+        !pageElements.menu.printerOtherSettings.classList.contains("notyet")
+      ) {
+        pageElements.menu.printerOtherSettings.classList.add("notyet");
```
If the printer goes offline whilst your on this page, this will make the UI react.

NotExpectedYet reviewed (commented):
Annotated.

davidzwa commented:
> Done the redirect mock, but not a clue with mocking passport.js at moment will look later when I have more time.

I wouldnt mock passport. I'd mock the `ensureCurrentUserAndGroup` and `ensureAuthenticated`

NotExpectedYet commented:
I've updated the 400 test and also added the 500 test. 

Is it worth doing the test for Runner at the moment? I had a quick go just mocking the 1 function I used and it was trying to do the whole init() function and I got a bit lost after that aha.

davidzwa commented:
API endpoints should not return 500. Thats a code for bugs or things that the developer should fix on the Octofarm server itself. 
400 series is for bad input or problems.

NotExpectedYet commented:
> API endpoints should not return 500. Thats a code for bugs or things that the developer should fix on the Octofarm server itself.
> 400 series is for bad input or problems.

It is a bug server side if that fails

NotExpectedYet commented:
There's a whole class of codes for Sever error, which is what this is catching so not sure what your on about with "API's shouldn't not return 500". 

The 5xx (Server Error) class of status code indicates that the server is aware that it has erred or is incapable of performing the requested method

davidzwa commented:
Yeah in hindsight my comment was a bit weird. I could explain how I come to think of 5xx, but its subjective probably. Lets move on haha

NotExpectedYet commented:
@davidzwa sound aha if you have reasoning I'm happy to hear it but no drama either way. XD

NotExpectedYet commented:
Is up to you regarding the tests. It's good to go apart from that I believe.

There's no output from the additional function I wrote in state, and it grabs stuff from OctoPrint so gather that's a larger task to test?

NotExpectedYet commented:
@davidzwa I've moved the refresh button to the side bar above save now. Was there some bugs you mentioned?

I fixed an issue that's happens when the port preference is null, it would incorrectly show the "Port not available" sign when there is no port preference. Found it because when you turn virtual printers on, the port preference is wiped on OctoPrint causing the refresh on OctoFarm to add a null in the port list. I don't see this been an issue in real world application, OctoFarm could not update the port preference if it detects a null from OctoPrint & OctoFarm has a saved value. I've not noticed OctoPrint wiping it in any other situation, or just magically forgetting like was mentioned in the discord. The preference get's saved in config.yaml and I tested an OctoPrint restart and it persisted.

davidzwa commented on client_src/js/lib/octofarm_client.js:
```diff
@@ -0,0 +1,42 @@
+        `Malformed request! Status: ${post.status} - ${post.statusText}`
+      );
+    } else if (post.status === 503) {
+      throw new Error("Error contacting server, is it alive?");
```
Are we ready for these errors to be thrown? Javascript breaks and quits on the first errors it sees.

NotExpectedYet replied:
I try catch everywhere I call that. And plan to do so for the rest I migrate over.  I'd love a global error handler but it's not something I'd say was worth our time pending V2. Or, could be done in another PR as I do want to migrate the client over to this file but the codes a reet spaghetti junction aha.

NotExpectedYet replied:
I gather I can move the try and catch inside this file and handle it there? We would have a global OctoFarm connection error handler then?

NotExpectedYet replied:
Wouldn't need to repeat it around the rest of the files either. Let me know will update it.

NotExpectedYet replied:
I always think of throw new Error, if caught to continue with the rest of the code. If this response is from the server then it's considered an error to me. Should this not be the case?

davidzwa replied:
All Im saying is that I spotted new code with an explicit `throw new Error(...)` which if thrown could break the UI.
If you are certain that we try-catch everything, sure no prob. I dont oversee that.

Oh note that code 503 and 400 are rare right now (400 is rarely thrown because we dont have input validation everywhere, its why I proposed using `npm-input-validation` as shown in octoprint-companion PR). 
Codes 0, 500, 404 are more common and I dont see all of those covered.

Oh and why an else on 204 if it doesnt do anything but return?

NotExpectedYet replied:
There's an else on 204 because I use 204 as a confirmation response from the server where no return data is needed. The return the function to let the rest skip which is at the bottom converting anything else to a JSON object. There would be nothing to convert if it didn't return and stop the .JSON.

davidzwa commented on server_src/routes/printers.js:
```diff
@@ -125,25 +127,26 @@ router.get("/groups", ensureAuthenticated, async (req, res) => {
+  const id = req.body.i;
+  if (!id) {
+    logger.error("Printer Settings: No ID key was provided");
+    res.statusMessage = "No ID key was provided";
```
I dont think this statusMessage scales. Reason is that error response often is more than just a message. It could become a code, a context object, a message - more than just a string. So, I foresee a moment where we'd have to revert this in future and that's why it doesnt work for me.

`throw "message"`
`throw new Error("error message")`
`throw new ValidationError(validationErrorInput)`
`throw new OctoPrintError(octoprintParseError)`
... others

All of these dont fit in statusMessage string.

davidzwa replied:
Oh btw, that does not stop this PR. It is just that I would not like to do this globally. That's all im sayin

NotExpectedYet replied:
I'm not using the status message for those eventualities there though? It's literally just saying "dodo, you need to provide a printer Id for this to work". Errors are caught below and didn't fire the status message? Can't remember. Anyway I think a global error handler should be tackled in another PR and update to use it. Would be good V2 prep work no?

davidzwa replied:
If your intention is to return input validation result, I propose to use `npm-input-validator` and return that result **in the body** + **400 code**. Much shorter code-wise, much clearer for another developer and much easier for the client to recognize and work with. This abstraction comes with so much benefit!
Here's how the input validation rules can look:
```
const checkInputRules = {
  persistenceUuid: "required",
  deviceUuid: "required",
  host: "string",
  port: "required|numeric",
  docker: "required|boolean",
  allowCrossOrigin: "required|boolean"
};
```
and this is the response if I leave out properties on purpose:
```
{
    "deviceUuid": {
        "message": "The device uuid field is mandatory.",
        "rule": "required"
    },
    "port": {
        "message": "The port field is mandatory.",
        "rule": "required"
    },
    "docker": {
        "message": "The docker field is mandatory.",
        "rule": "required"
    }
}
```
I had to do nothing for it (except for calling the validation and returning it if failed)! And because the rules are constant strings, we dont have to guess what is the reason for failure on the client side. Its explicit and reliable.

If input validation is what you're looking to abstract, I hope you like the method I propose above. This would make porting API code over to V2 an absolute breeze! 💨 💨💨

davidzwa reviewed (changes requested):
So to summarize
- This PR is epic
- Youre doing such nice work!
- I get to learn so much right now, this is so handy for thinking about stuff for V2
- Merge solved
- Fixed the tests
- I applied the list/getById change
- I applied prettier to some client files (only partially, more to be done)
- the error throwing in client `octofarm.js` code => risk analysis (maybe test with some test errors server-side, see how it looks?) I think this should be postponed for more breathing room as client side is a risk factor.
- After settings have been updated we see 3 messages (maybe a bit over the top)
![image](https://user-images.githubusercontent.com/6005355/119569562-302b4280-bdaf-11eb-8162-71f06e27e2ee.png)
- After saving the settings, in case we had the **red** portPreference problem and tried to correct for it, the red color does not disappear although the situation is fixed.
- Saving/refreshing is quite slow. Are there concessions possible to speed this up? Could it be that state.js is more work than necessary? (We can follow-up with a PR ofc)
- think about statusMessage, see my feedback above and think if this scales for the idea you have in mind in case we need to add error context (f.e. API validation errors or OctoPrint remote API errors). Please communicate what your plans are 👍🏼

NotExpectedYet commented:
> So to summarize
> 
> * This PR is epic
> * Youre doing such nice work!
> * I get to learn so much right now, this is so handy for thinking about stuff for V2
> * Merge solved
> * Fixed the tests
> * I applied the list/getById change
> * I applied prettier to some client files (only partially, more to be done)
> * the error throwing in client `octofarm.js` code => risk analysis (maybe test with some test errors server-side, see how it looks?) I think this should be postponed for more breathing room as client side is a risk factor.

It was tested but commented above, regarding moving the try/catch then we can deal with it in the file. I'm not depended on using throw new error, would just like to know alternatives and how better to deal with it. 

> * After settings have been updated we see 3 messages (maybe a bit over the top)
>   ![image](https://user-images.githubusercontent.com/6005355/119569562-302b4280-bdaf-11eb-8162-71f06e27e2ee.png)

Those are because of the 3 calls the saving settings actually makes. It contacts OctoFarm, it contacts OctoPrint's API twice. I'll see about putting it all into a single notification and doing warning colour if one fails instead. 

> * After saving the settings, in case we had the **red** portPreference problem and tried to correct for it, the red color does not disappear although the situation is fixed.

Hmm that went on mine, will double check. Any reproduction? as I was using virtual printers for testing. 

> * Saving/refreshing is quite slow. Are there concessions possible to speed this up? Could it be that state.js is more work than necessary? (We can follow-up with a PR ofc)

Not from my experience here. Saving was rapid and so was refresh. Is this a windows vs linux thing again? 

> * think about statusMessage, see my feedback above and think if this scales for the idea you have in mind in case we need to add error context (f.e. API validation errors or OctoPrint remote API errors). Please communicate what your plans are, my opinion is flexible but my code-smell nose is not 👍🏼

Replied above.

davidzwa commented:
> > * the error throwing in client `octofarm.js` code => risk analysis (maybe test with some test errors server-side, see how it looks?) I think this should be postponed for more breathing room as client side is a risk factor.
> 
> It was tested but commented above, regarding moving the try/catch then we can deal with it in the file. I'm not depended on using throw new error, would just like to know alternatives and how better to deal with it.

Responded above

> Those are because of the 3 calls the saving settings actually makes. It contacts OctoFarm, it contacts OctoPrint's API twice. I'll see about putting it all into a single notification and doing warning colour if one fails instead.

That's nice

> Hmm that went on mine, will double check. Any reproduction? as I was using virtual printers for testing.

I used my Pi4 + prusa mini to reproduce as this is a portPref/udev problem and virtual does not work for that? This is a physical problem.
1) I turned on my printer (ttyACM0, not VIRTUAL)
2) connected to it in OP + saved
2b) checked config: ttyACM0 saved
3) powered off the printer 
4) I refreshed OctoPrint to check the 'usb serial not found blabla'
5) I checked OF, and went to the printer dialog (it was red: expected)
6) I rectified it in the dialog and pressed save.
7) Saw that the red color didnt dissappear. If Im correct I also pressed the refresh to make sure.
7b) I checked OP + config: problem was gone there and serial was set back to VIRTUAL.
8) I reopened the dialog and the red color was gone, so most probably just UI logic to change the color missing or not working as expected.

So I think this will occur for your situation as well. Please try again.

> > * Saving/refreshing is quite slow. Are there concessions possible to speed this up? Could it be that state.js is more work than necessary? (We can follow-up with a PR ofc)
> Not from my experience here. Saving was rapid and so was refresh. Is this a windows vs linux thing again?

My prusa/Pi4 is running on the network, so I think this is just delay because of calling multiple OP API endpoints in a row synchronously. My question is can we optimize this? If I save printer settings, I dont want the UI to wait for regenerating cache 
 or calling to reload profiles, since we have SSE for that (and hopefully WS later). Look the reason Im hammering on this is because it seems that this is suboptimal and the core business of OF to be as snappy as possible.

This has nothing to do with Windows. That was not a problem. The reason for the 5 sec time was because SystemRunner just does a lot of work for 

> 
> > * think about statusMessage, see my feedback above and think if this scales for the idea you have in mind in case we need to add error context (f.e. API validation errors or OctoPrint remote API errors). Please communicate what your plans are, my opinion is flexible but my code-smell nose is not 👍🏼
> 
> Replied above.

I hope you like the `npm-input-validator` approach, now that I showed some examples for it. This can be followed up in another PR if you wish. This work is the biggest gap to fill to hop onto the V2 train btw, especially because I see how the cleaners and state work now. Well worth our time to discuss

NotExpectedYet commented:
Ah, I didn't plan for that eventuality. I went to octoprint and resolved the issue then rescanned. I didn't sort in OctoFarm because I presumed they'd have to go there. The refresh cleared it for me though. I gather all that needs is a refresh of the page on save as that isn't done currently.

NotExpectedYet commented:
@davidzwa Made the changes, and I've removed most of those calls from state. One of the checks isn't firing for reScaning the OctoPrint instance when you updated the URL that time but think that function needs cleaning up in another PR. The whole difference function checks need looking at tbh. Anyway, now saves. I decided not to put the try/catch inside the function for octofarm_client as I think the response should be handled higher up. I've also captured a screen recording of exactly how fast the older method of refresh/saving was for me so I don't think it's the multiple calls to OP that's causing the lagged response on your end unless I'm misunderstanding you, to me that would be reproducible here no? 

![Peek 2021-05-29 07-38](https://user-images.githubusercontent.com/61137988/120060729-540db300-c051-11eb-85ea-2e1fca0fc2d6.gif)

davidzwa commented on server_src/runners/state.js:
```diff
@@ -3473,26 +3479,7 @@ class Runner {
-      await Runner.getOctoPrintSystenInfo(settings.printer.index);
-      Runner.getOctoPrintSystenInfo(settings.printer.index);
-      Runner.getUpdates(settings.printer.index);
-      Runner.getPluginList(settings.printer.index);
```
Mark this in your memory that if there are problems with settings we can add any of it back. I dont think we will have to, but just note.

davidzwa reviewed (approved):
YAAAAS
